### PR TITLE
Add Rouchon solver of order 1 for SME

### PR DESCRIPTION
**Pull request by pierreguilmin**
Here's a little cheatsheet to help checking the solver, it's actually very similar to Rouchon of order 1 for `mesolve`.

Screenshot 2023-09-08 at 16 41 48

Note that purely dissipative processes (jump operators not monitored, named $T$ in the code) have $\eta_k=0$, and we can thus compute terms that depend on $\eta_k$ only for monitored jump operators (named $V$ in the code).

### CODE REVIEW

File: dynamiqs/smesolve/sme_solver.py [hunk 1]
```diff
@@ -15,24 +15,27 @@ def __init__(
...
-        self.generator = generator
         self.t_meas = t_meas  # (len(t_meas))
-
         super().__init__(*args, jump_ops=jump_ops)
 
+        # split jump operators between purely dissipative (eta = 0) and
+        # monitored (eta != 0)
+        mask = etas == 0.0
+        self.T = jump_ops[mask]  # (len(T), n, n) purely dissipative
+        self.V = jump_ops[~mask]  # (len(V), n, n) monitored
```
Comment (gautierronan) [inline dynamiqs/smesolve/sme_solver.py:29]: Are these notations from somewhere? IMO they suggest kinetic (T) and potential (V) energy terms. Could we use something else? Maybe `self.J` (jump) or `self.C` (collapse) and `self.M` (measured/monitored)?

File: dynamiqs/smesolve/sme_solver.py [hunk 2]
```diff
@@ -15,24 +15,27 @@ def __init__(
...
         etas: Tensor,
         generator: torch.Generator,
         t_meas: Tensor,
     ):
-        self.meas_ops = meas_ops  # (len(meas_ops), n, n)
-        self.etas = etas  # (len(meas_ops))
-        self.generator = generator
         self.t_meas = t_meas  # (len(t_meas))
-
         super().__init__(*args, jump_ops=jump_ops)
```
Comment (gautierronan) [inline dynamiqs/smesolve/sme_solver.py:23]: Can you rename `jump_ops` to `L` like you did in mesolve (I think?)?

File: dynamiqs/smesolve/sme_solver.py [hunk 3]
```diff
@@ -79,37 +82,40 @@ def sample_wiener(self, dt: float) -> Tensor:
...
     @cache
-    def Lp(self, rho: Tensor) -> Tensor:
-        # rho: (b_H, b_rho, ntrajs, n, n) -> (b_H, b_rho, ntrajs, n, n)
-        # L @ rho + rho @ Ldag
-        L_rho = torch.einsum('bij,...jk->...bik', self.meas_ops, rho)
-        return L_rho + L_rho.mH
+    def Vp(self, rho: Tensor) -> Tensor:
+        # rho: (b_H, b_rho, ntrajs, n, n) -> (b_H, b_rho, ntrajs, len(V), n, n)
+        # V @ rho + rho @ Vdag
+        V_rho = torch.einsum('bij,...jk->...bik', self.V, rho)
```
Comment (gautierronan) [inline dynamiqs/smesolve/sme_solver.py:88]: Probably faster to use something else than einsum when possible. Here I'm guessing a simple `@` is best.

**Review by gautierronan — CHANGES_REQUESTED** [review]
Nice. Did not check the formulas, so may need a double check

File: dynamiqs/smesolve/sme_solver.py (hunk 2 above)
Comment (pierreguilmin) [reply dynamiqs/smesolve/sme_solver.py:23]: `SMESolver` inherits from `MESolver` so it's done automatically.

File: dynamiqs/smesolve/sme_solver.py (hunk 3 above)
Comment (pierreguilmin) [reply dynamiqs/smesolve/sme_solver.py:88]: `@` doesn't work (`self.V` has shape `(len(V), n, n)` and `rho` has shape `(..., n, n)`).

File: dynamiqs/smesolve/sme_solver.py (hunk 1 above)
Comment (pierreguilmin) [reply dynamiqs/smesolve/sme_solver.py:29]: `self.M` conflicts with `Rouchon` solver `self.M0/self.M1s` and with the measurement back-action superoperator `\mathcal{M}`. 😕 I picked `Lc` (for collapse) and `Lm` (for monitored), tell me if that works for you.

Comment (gautierronan) [reply dynamiqs/smesolve/sme_solver.py]: Perfect !

**Review by gautierronan — APPROVED** [review]
Great job !

**Pull request merged**